relational-query/src/Database/Relational/Monad/Trans/Join.hs

```diff
@@ -61,14 +63,15 @@ join' :: Monad m => m a -> QueryJoin m a
 join' = lift
 
 -- | Unsafely update join product context.
-updateContext :: Monad m => (JoinContext -> JoinContext) -> QueryJoin m ()
-updateContext =  QueryJoin . modify
+updateContext :: Monad m => PlaceholderOffsets -> (JoinContext -> JoinContext) -> QueryJoin m ()
+updateContext phs f =
+  QueryJoin $ modify (mapWithPlaceholderOffsets (f *** (phs <>)))
```

Comment:
I found a bug related to this file: there are cases where the predicate of `ON` clause are not printed even if its attached `PlaceholderOffsets` is appended.